### ISSUE TITLE
SPDK v18.07 can't detect all my SSD disks

### ISSUE THREAD

**Issue by WilsonkcLo** [spdk/spdk#477]
I install SPDK v18.07 on my system with 18x SSD disk.  But it only detects two disks after I run the setup.sh.  If I run the "setup.sh reset", it shows all 18x nvme disks.  I run the "pref" command on the detected disk and the command works fine.

## Expected Behavior
Shows all disks after running the setup.sh

## Current Behavior
Only shows two disks out of 18x disks.  Log file is below -
***** Before run the setup.sh
amd@amd-s1:~/spdk$ ./scripts/setup.sh status
Hugepages
node     hugesize     free /  total
node0   1048576kB        0 /      0
node0      2048kB      512 /    512
node1   1048576kB        0 /      0
node1      2048kB      512 /    512
node2   1048576kB        0 /      0
node2      2048kB      512 /    512
node3   1048576kB        0 /      0
node3      2048kB      512 /    512
NVMe devices
BDF             Numa Node       Driver name             Device name
0000:02:00.0    0               nvme                    nvme0
0000:03:00.0    0               nvme                    nvme1
0000:04:00.0    0               nvme                    nvme2
0000:05:00.0    0               nvme                    nvme3
0000:41:00.0    0               nvme                    nvme16
0000:42:00.0    0               nvme                    nvme17
0000:47:00.0    0               nvme                    nvme4
0000:48:00.0    0               nvme                    nvme5
0000:49:00.0    0               nvme                    nvme6
0000:4a:00.0    0               nvme                    nvme7
0000:81:00.0    0               nvme                    nvme8
0000:82:00.0    0               nvme                    nvme9
0000:83:00.0    0               nvme                    nvme10
0000:84:00.0    0               nvme                    nvme11
0000:c1:00.0    0               nvme                    nvme12
0000:c2:00.0    0               nvme                    nvme13
0000:c3:00.0    0               nvme                    nvme14
0000:c4:00.0    0               nvme                    nvme15
I/OAT DMA
BDF             Numa Node       Driver Name
virtio
BDF             Numa Node       Driver Name             Device Name

***** Result of setup.sh
amd@amd-s1:~/spdk$ sudo ./scripts/setup.sh
0000:02:00.0 (144d a808): nvme -> uio_pci_generic
0000:03:00.0 (144d a808): nvme -> uio_pci_generic
0000:04:00.0 (144d a808): nvme -> uio_pci_generic
0000:05:00.0 (144d a808): nvme -> uio_pci_generic
0000:41:00.0 (144d a808): nvme -> uio_pci_generic
0000:42:00.0 (144d a808): nvme -> uio_pci_generic
0000:47:00.0 (144d a808): nvme -> uio_pci_generic
0000:48:00.0 (144d a808): nvme -> uio_pci_generic
0000:49:00.0 (144d a808): nvme -> uio_pci_generic
0000:4a:00.0 (144d a808): nvme -> uio_pci_generic
0000:81:00.0 (144d a808): nvme -> uio_pci_generic
0000:82:00.0 (144d a808): nvme -> uio_pci_generic
0000:83:00.0 (144d a808): nvme -> uio_pci_generic
0000:84:00.0 (144d a808): nvme -> uio_pci_generic
0000:c1:00.0 (144d a808): nvme -> uio_pci_generic
0000:c2:00.0 (144d a808): nvme -> uio_pci_generic
0000:c3:00.0 (144d a808): nvme -> uio_pci_generic
0000:c4:00.0 (144d a808): nvme -> uio_pci_generic

***** SPDK status after running setup.sh
amd@amd-s1:~/spdk$ ./scripts/setup.sh status
Hugepages
node     hugesize     free /  total
node0   1048576kB        0 /      0
node0      2048kB      256 /    256
node1   1048576kB        0 /      0
node1      2048kB      256 /    256
node2   1048576kB        0 /      0
node2      2048kB      256 /    256
node3   1048576kB        0 /      0
node3      2048kB      256 /    256
NVMe devices
BDF             Numa Node       Driver name             Device name
0000:02:00.0    0                               -
0000:03:00.0    0                               -
0000:04:00.0    0                               -
0000:05:00.0    0                               -
0000:41:00.0    0               uio_pci_generic         -
0000:42:00.0    0               uio_pci_generic         -
0000:47:00.0    0                               -
0000:48:00.0    0                               -
0000:49:00.0    0                               -
0000:4a:00.0    0                               -
0000:81:00.0    0                               -
0000:82:00.0    0                               -
0000:83:00.0    0                               -
0000:84:00.0    0                               -
0000:c1:00.0    0                               -
0000:c2:00.0    0                               -
0000:c3:00.0    0                               -
0000:c4:00.0    0                               -
I/OAT DMA
BDF             Numa Node       Driver Name
virtio
BDF             Numa Node       Driver Name             Device Name

## Steps to Reproduce
1. git clone https://github.com/spdk/spdk.git
2. cd spdk/
3. git checkout v18.07
4. git submodule update --init
5. ./configure
6. make

## Context (Environment including OS version, SPDK version, etc.)
1. Linux amd-s1 4.15.0-36-generic #39-Ubuntu SMP 
2. SPDK v18.07

**Comment by jimharris** [comment]
Hi,

I have a few questions/requests:

Please provide anything you see in dmesg after you've run these setup.sh commands.
Please also try running the setup.sh script again, but after modifying it to remove the redirections to /dev/null in linux_bind_driver().  That may provide some clues.
Can you provide some more details on the platform?
"I run the "pref" command on the detected disk and the command works fine." -> do you mean the nvme/perf command?  and which detected disk?

Thanks,

-Jim

**Comment by WilsonkcLo** [comment]
I can only run the command with PCIe address "0000:41:00.0" or "0000:42:00.0", other address will get undetected physical address failure (command is below)
~/spdk/examples/nvme/perf/perf -q 128 -o 4096 -w randread -r 'trtype:PCIe traddr:0000:41:00.0' -t 300

If I remove the redirection in linux_bind_driver(), the dmesg is below 
amd@amd-s1:~/spdk$ sudo scripts/setup.sh
0000:02:00.0 (144d a808): nvme -> uio_pci_generic
scripts/setup.sh: line 109: echo: write error: No such device

**Comment by jimharris** [comment]
Hi @WilsonkcLo,

setup.sh line 109 on v18.07 does not look like a write operation.  Is it possible you have locally modified setup.sh?

Some more details on the platform would also be appreciated.

Thanks,

-Jim

**Comment by xuhuagen** [comment]
My environment: 
CentOS Linux release 7.5.1804 (Core)
kernel: 4.17.1
I git checkout v18.07, run the "pref" command and all my SSD disks can be detected.

**Comment by WilsonkcLo** [comment]
Hi,
The setup.sh is the original file.  I would like to confirm the modification of setup.sh you suggested. I remove the "2 >/dev/null || true" in line 100, 106 and 107.  For example, 
The line 100 is changed from - echo "$ven_dev_id" > "/sys/bus/pci/devices/$bdf/driver/remove_id" 2> /dev/null || true , to echo "$ven_dev_id" > "/sys/bus/pci/devices/$bdf/driver/remove_id"

Does my modification correct?  May I know what platform information you need to investigate the issue.  If you would like to have more information, please let me know.

After I run the setup.sh (original version), I can't see any disk in lsblk command.  So, I assume all disks are not visible by kernel any more and all /dev/nvme* node are removed.

Before running the setup.sh
amd@amd-s1:~/spdk$ lsblk
NAME     MAJ:MIN RM   SIZE RO TYPE MOUNTPOINT
loop1      7:1    0  87.9M  1 loop /snap/core/5328
loop2      7:2    0  87.9M  1 loop /snap/core/5548
loop3      7:3    0  87.9M  1 loop /snap/core/5662
sda        8:0    0 465.8G  0 disk
├─sda1     8:1    0   512M  0 part /boot/efi
└─sda2     8:2    0 465.3G  0 part /
sr0       11:0    1  1024M  0 rom
sr1       11:1    1  1024M  0 rom
sr2       11:2    1  1024M  0 rom
sr3       11:3    1  1024M  0 rom
nvme1n1  259:0    0   3.5T  0 disk
nvme0n1  259:1    0   3.5T  0 disk
nvme6n1  259:2    0   3.5T  0 disk
nvme2n1  259:3    0   3.5T  0 disk
nvme5n1  259:4    0   3.5T  0 disk
nvme4n1  259:5    0   3.5T  0 disk
nvme7n1  259:6    0   3.5T  0 disk
nvme8n1  259:7    0   3.5T  0 disk
nvme3n1  259:8    0   3.5T  0 disk
nvme9n1  259:9    0   3.5T  0 disk
nvme10n1 259:10   0   3.5T  0 disk
nvme11n1 259:11   0   3.5T  0 disk
nvme12n1 259:12   0   3.5T  0 disk
nvme14n1 259:13   0   3.5T  0 disk
nvme13n1 259:14   0   3.5T  0 disk
nvme15n1 259:15   0   3.5T  0 disk
nvme16n1 259:16   0   3.5T  0 disk
nvme17n1 259:17   0   3.5T  0 disk

Command: lspci
amd@amd-s1:~/spdk$  lspci -vvv |grep a808
02:00.0 Non-Volatile memory controller: Samsung Electronics Co Ltd Device a808 (prog-if 02 [NVM Express])
03:00.0 Non-Volatile memory controller: Samsung Electronics Co Ltd Device a808 (prog-if 02 [NVM Express])
04:00.0 Non-Volatile memory controller: Samsung Electronics Co Ltd Device a808 (prog-if 02 [NVM Express])
05:00.0 Non-Volatile memory controller: Samsung Electronics Co Ltd Device a808 (prog-if 02 [NVM Express])
41:00.0 Non-Volatile memory controller: Samsung Electronics Co Ltd Device a808 (prog-if 02 [NVM Express])
42:00.0 Non-Volatile memory controller: Samsung Electronics Co Ltd Device a808 (prog-if 02 [NVM Express])
47:00.0 Non-Volatile memory controller: Samsung Electronics Co Ltd Device a808 (prog-if 02 [NVM Express])
48:00.0 Non-Volatile memory controller: Samsung Electronics Co Ltd Device a808 (prog-if 02 [NVM Express])
49:00.0 Non-Volatile memory controller: Samsung Electronics Co Ltd Device a808 (prog-if 02 [NVM Express])
4a:00.0 Non-Volatile memory controller: Samsung Electronics Co Ltd Device a808 (prog-if 02 [NVM Express])
81:00.0 Non-Volatile memory controller: Samsung Electronics Co Ltd Device a808 (prog-if 02 [NVM Express])
82:00.0 Non-Volatile memory controller: Samsung Electronics Co Ltd Device a808 (prog-if 02 [NVM Express])
83:00.0 Non-Volatile memory controller: Samsung Electronics Co Ltd Device a808 (prog-if 02 [NVM Express])
84:00.0 Non-Volatile memory controller: Samsung Electronics Co Ltd Device a808 (prog-if 02 [NVM Express])
c1:00.0 Non-Volatile memory controller: Samsung Electronics Co Ltd Device a808 (prog-if 02 [NVM Express])
c2:00.0 Non-Volatile memory controller: Samsung Electronics Co Ltd Device a808 (prog-if 02 [NVM Express])
c3:00.0 Non-Volatile memory controller: Samsung Electronics Co Ltd Device a808 (prog-if 02 [NVM Express])
c4:00.0 Non-Volatile memory controller: Samsung Electronics Co Ltd Device a808 (prog-if 02 [NVM Express])

**Comment by WilsonkcLo** [comment]
Detail view of PCI bus
amd@amd-s1:~/spdk$ lspci -vt
-+-[0000:c0]-+-01.1-[c1]----00.0  Samsung Electronics Co Ltd Device a808
 |           +-01.2-[c2]----00.0  Samsung Electronics Co Ltd Device a808
 |           +-01.3-[c3]----00.0  Samsung Electronics Co Ltd Device a808
 |           +-01.4-[c4]----00.0  Samsung Electronics Co Ltd Device a808
 |           +-03.1-[c5]----00.0  Mellanox Technologies MT27700 Family [ConnectX-4]
 +-[0000:80]-+-01.1-[81]----00.0  Samsung Electronics Co Ltd Device a808
 |           +-01.2-[82]----00.0  Samsung Electronics Co Ltd Device a808
 |           +-01.3-[83]----00.0  Samsung Electronics Co Ltd Device a808
 |           +-01.4-[84]----00.0  Samsung Electronics Co Ltd Device a808
 |           +-03.1-[85]----00.0  Mellanox Technologies MT27800 Family [ConnectX-5]
 +-[0000:40]-+-01.2-[41]----00.0  Samsung Electronics Co Ltd Device a808
 |           +-01.3-[42]----00.0  Samsung Electronics Co Ltd Device a808
 |           +-01.4-[43]----00.0  ASMedia Technology Inc. ASM1062 Serial ATA Controller
 |           +-01.5-[44]----00.0  ASMedia Technology Inc. ASM1062 Serial ATA Controller
 |           +-01.6-[45-46]----00.0-[46]----00.0  ASPEED Technology, Inc. ASPEED Graphics Family
 |           +-03.1-[47]----00.0  Samsung Electronics Co Ltd Device a808
 |           +-03.2-[48]----00.0  Samsung Electronics Co Ltd Device a808
 |           +-03.3-[49]----00.0  Samsung Electronics Co Ltd Device a808
 |           +-03.4-[4a]----00.0  Samsung Electronics Co Ltd Device a808
 \-[0000:00]-+-01.1-[01]----00.0  Mellanox Technologies MT27700 Family [ConnectX-4]
             +-03.1-[02]----00.0  Samsung Electronics Co Ltd Device a808
             +-03.2-[03]----00.0  Samsung Electronics Co Ltd Device a808
             +-03.3-[04]----00.0  Samsung Electronics Co Ltd Device a808
             +-03.4-[05]----00.0  Samsung Electronics Co Ltd Device a808

**Comment by WilsonkcLo** [comment]
My software information is 
amd@amd-s1:~/spdk$ cat /etc/*release
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=18.04
DISTRIB_CODENAME=bionic
DISTRIB_DESCRIPTION="Ubuntu 18.04.1 LTS"
NAME="Ubuntu"
VERSION="18.04.1 LTS (Bionic Beaver)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.04.1 LTS"
VERSION_ID="18.04"

**Comment by WilsonkcLo** [comment]
Hi,
I found the issue is in line 100.  After running the setup.sh, the "driver" directory is missing under PCIe 0000:02:00.0 and others (before running the script, the directory is there) .  The "driver" directory only shows in "0000:41:00.0" and "0000:42:00.0" which are SPDK detected disk.  And it causes the write error.

amd@amd-s1:~$ **Before running setup.sh**
Before: command not found
amd@amd-s1:~$ sudo ls /sys/bus/pci/devices/0000:02:00.0
broken_parity_status      current_link_speed  dma_mask_bits    irq             max_link_width  numa_node  remove    resource0         subsystem_vendor
class                     current_link_width  **driver**           local_cpulist   modalias        nvme       rescan    revision          uevent
config                    d3cold_allowed      driver_override  local_cpus      msi_bus         pools      reset     subsystem         vendor
consistent_dma_mask_bits  device              enable           max_link_speed  msi_irqs        power      resource  subsystem_device
amd@amd-s1:~$ sudo ls /sys/bus/pci/devices/0000:41:00.0
broken_parity_status      current_link_speed  dma_mask_bits    irq             max_link_width  numa_node  remove    resource0  subsystem_device
class                     current_link_width  **driver**           local_cpulist   modalias        nvme       rescan    revision   subsystem_vendor
config                    d3cold_allowed      driver_override  local_cpus      msi_bus         pools      reset     rom        uevent
consistent_dma_mask_bits  device              enable           max_link_speed  msi_irqs        power      resource  subsystem  vendor
amd@amd-s1:~$ **After running setup.sh**
After: command not found
amd@amd-s1:~$ **sudo ls /sys/bus/pci/devices/0000:02:00.0**
broken_parity_status      current_link_speed  dma_mask_bits    local_cpulist   modalias   remove    resource0         subsystem_vendor
class                     current_link_width  driver_override  local_cpus      msi_bus    rescan    revision          uevent
config                    d3cold_allowed      enable           max_link_speed  numa_node  reset     subsystem         vendor
consistent_dma_mask_bits  device              irq              max_link_width  power      resource  subsystem_device
amd@amd-s1:~$ **sudo ls /sys/bus/pci/devices/0000:41:00.0**
broken_parity_status      current_link_speed  dma_mask_bits    irq             max_link_width  power   resource   subsystem         uio
class                     current_link_width  **driver**           local_cpulist   modalias        remove  resource0  subsystem_device  vendor
config                    d3cold_allowed      driver_override  local_cpus      msi_bus         rescan  revision   subsystem_vendor
consistent_dma_mask_bits  device              enable           max_link_speed  numa_node       reset   rom        uevent
amd@amd-s1:~$ sudo ls /sys/bus/pci/devices/0000:42:00.0
broken_parity_status      current_link_speed  dma_mask_bits    irq             max_link_width  power   resource   subsystem         uio
class                     current_link_width  driver           local_cpulist   modalias        remove  resource0  subsystem_device  vendor
config                    d3cold_allowed      driver_override  local_cpus      msi_bus         rescan  revision   subsystem_vendor
consistent_dma_mask_bits  device              enable           max_link_speed  numa_node       reset   rom        uevent
amd@amd-s1:~$

**Comment by jimharris** [comment]
Note: this looks like a duplicate of https://github.com/spdk/spdk/issues/273.

**Comment by jimharris** [comment]
Line 100 does an echo of $ven_dev_id to .../$bdev/driver/remove_id - this is the line that's causing the write error?  Can you double-check?

**Comment by WilsonkcLo** [comment]
Yes, I confirm this line cause the error message - 
echo "$ven_dev_id" > "/sys/bus/pci/devices/$bdf/driver/remove_id" || true

I added an echo before and after this line.  Thanks

**Comment by jimharris** [comment]
Interesting.

So this means that before running setup.sh, the $bdf/driver directory is there, but it doesn't have a remove_id file.  Could you confirm?

Could you also try commenting out the "set -e" at the top of the file, and re-run?  set -e is causing the script to exit after the first error since we removed the "|| true".

Thanks.

**Comment by jimharris** [comment]
Line 100 is before we've tried to attach the device to uio_pci_generic.  The driver directory should still be there at this point.  At line 100, we are trying to unbind the device from the kernel nvme driver first.  So I'd like an ls of the driver directory before the script is run to confirm that remove_id is not there.

**Comment by WilsonkcLo** [comment]
It is before the script to run
amd@amd-s1:~/spdk$ ls -lL /sys/bus/pci/devices/0000:02:00.0/driver/
total 0
drwxr-xr-x 5 root root    0 Oct 16 18:49 0000:02:00.0
drwxr-xr-x 5 root root    0 Oct 16 18:49 0000:03:00.0
drwxr-xr-x 5 root root    0 Oct 16 18:49 0000:04:00.0
drwxr-xr-x 5 root root    0 Oct 16 18:49 0000:05:00.0
drwxr-xr-x 5 root root    0 Oct 16 18:49 0000:41:00.0
drwxr-xr-x 5 root root    0 Oct 25 18:04 0000:42:00.0
drwxr-xr-x 5 root root    0 Oct 16 18:49 0000:47:00.0
drwxr-xr-x 5 root root    0 Oct 16 18:49 0000:48:00.0
drwxr-xr-x 5 root root    0 Oct 16 18:49 0000:49:00.0
drwxr-xr-x 5 root root    0 Oct 16 18:49 0000:4a:00.0
drwxr-xr-x 5 root root    0 Oct 16 18:49 0000:81:00.0
drwxr-xr-x 5 root root    0 Oct 16 18:49 0000:82:00.0
drwxr-xr-x 5 root root    0 Oct 16 18:49 0000:83:00.0
drwxr-xr-x 5 root root    0 Oct 16 18:49 0000:84:00.0
drwxr-xr-x 5 root root    0 Oct 16 18:49 0000:c1:00.0
drwxr-xr-x 5 root root    0 Oct 16 18:49 0000:c2:00.0
drwxr-xr-x 5 root root    0 Oct 16 18:49 0000:c3:00.0
drwxr-xr-x 5 root root    0 Oct 16 18:49 0000:c4:00.0
--w------- 1 root root 4096 Oct 25 18:04 bind
drwxr-xr-x 7 root root    0 Oct 16 18:49 module
--w------- 1 root root 4096 Oct 25 18:04 new_id
--w------- 1 root root 4096 Oct 25 17:58 remove_id
--w------- 1 root root 4096 Oct 16 18:49 uevent
--w------- 1 root root 4096 Oct 25 17:58 unbind

****************************
After the script is run
****************************
amd@amd-s1:~/spdk$ ls -lL /sys/bus/pci/devices/0000:02:00.0/driver/
ls: cannot access '/sys/bus/pci/devices/0000:02:00.0/driver/': No such file or directory

**Comment by WilsonkcLo** [comment]
Looks like it failed to link the folder in all PCIe disk except "0000.41.00.0".
amd@amd-s1:~/spdk$ ls /sys/bus/pci/devices/0000:41:00.0/driver -l
lrwxrwxrwx 1 root root 0 Oct 25 18:36 /sys/bus/pci/devices/0000:41:00.0/driver -> ../../../../bus/pci/drivers/uio_pci_generic

**Comment by jimharris** [comment]
OK.  This remove_id error is not relevant then.  This device/vendor ID was not added as a new_id, so remove_id_store() is returning -ENODEV.

Could you try commenting out the "set -e", do a setup.sh reset, and then setup.sh again?  The log attachment you did earlier was perfect - you can do that instead of trying to paste the text inline.

Thanks,

-Jim

**Comment by WilsonkcLo** [comment]
I don't see any different after "set- e" is comment out.

**Comment by jimharris** [comment]
Sorry - I just realized there are a bunch of "set -e" throughout this script.  You'll need to comment out all of them.

**Comment by WilsonkcLo** [comment]
I don't see any additional output message after removing "set -e"

**Comment by WilsonkcLo** [comment]
During running the setup.sh, I find there is no error message on "0000:02:00.0", but "0000:41:00.0" has ..... However, setup.sh status shows "0000:41:00.0" is mounted.

amd@amd-s1:~/spdk$ sudo scripts/setup.sh
Before line 100
After line 100
0000:02:00.0 (144d a808): nvme -> uio_pci_generic
Before line 100
After line 100
0000:03:00.0 (144d a808): nvme -> uio_pci_generic
Before line 100
After line 100
0000:04:00.0 (144d a808): nvme -> uio_pci_generic
Before line 100
scripts/setup.sh: line 104: echo: write error: No such device
After line 100
0000:05:00.0 (144d a808): nvme -> uio_pci_generic
Before line 100
scripts/setup.sh: line 104: echo: write error: No such device
After line 100
0000:41:00.0 (144d a808): nvme -> uio_pci_generic
Before line 100
scripts/setup.sh: line 104: echo: write error: No such device
After line 100
0000:42:00.0 (144d a808): nvme -> uio_pci_generic
.....
.....
.....

**Comment by WilsonkcLo** [comment]
It is AMD x86_64.
I only set one entry in PCI_WHITELIST and skipping all others.  I found that the initialization of setup.sh has error, but reports success.  When I run the script, it reports "0000:02:00.0 (144d a808): nvme -> uio_pci_generic".  However, when I run the "reset", it reports "no driver".  The log is below -

amd@amd-s1:~/spdk$ sudo PCI_WHITELIST="0000:02:00.0" scripts/setup.sh
Before line 100
After line 100
0000:02:00.0 (144d a808): nvme -> uio_pci_generic
Skipping un-whitelisted NVMe controller nvme7n1 (0000:03:00.0)
Skipping un-whitelisted NVMe controller nvme6n1 (0000:04:00.0)
Skipping un-whitelisted NVMe controller nvme2n1 (0000:05:00.0)
......
......
amd@amd-s1:~/spdk$ sudo scripts/setup.sh status
......
......
NVMe devices
BDF             Numa Node       Driver name             Device name
0000:02:00.0    0                               -
0000:03:00.0    0               nvme                    nvme1
0000:04:00.0    0               nvme                    nvme2
0000:05:00.0    0               nvme                    nvme3
0000:41:00.0    0               nvme                    nvme16
.....
.....
I/OAT DMA
BDF             Numa Node       Driver Name
virtio
BDF             Numa Node       Driver Name             Device Name

amd@amd-s1:~/spdk$ sudo scripts/setup.sh reset
0000:02:00.0 (144d a808): no driver -> nvme      <=========
amd@amd-s1:~/spdk$

**Comment by jimharris** [comment]
Does dmesg show anything?  Any errors from the uio_pci_generic driver?

**Comment by jimharris** [comment]
This is certainly not an SPDK error at this point.  I think you'll need to do more instrumentation in the uio_pci_generic driver to see why it can't bind to most of these SSDs.  You may also try enabling the IOMMU and using vfio-pci instead.

I'm going to close #273 as a duplicate of this issue here, since this one has more analysis information.

**Comment by WilsonkcLo** [comment]
Thanks for your help.  Yes, I believe it is kernel driver issue.  See the dmesg after I run the script with one disk only.

[782797.940915] pcieport 0000:00:03.1: can't derive routing for PCI INT A
[782797.940920] uio_pci_generic 0000:02:00.0: PCI INT A: no GSI
[782797.941100] uio_pci_generic: probe of 0000:02:00.0 failed with error -107

**Comment by WilsonkcLo** [comment]
For your information, the issue is fixed by using igb_uio driver instead of uio_pci_generic.
